community/louvain_common.go

```diff
@@ -35,11 +36,194 @@ func Q(g graph.Graph, communities [][]graph.Node, resolution float64) float64 {
+// returned as the vector of weighted Q scores for each layer of the multiplex graph.
+// If communities is nil, the unclustered modularity score is returned.
+// If weights is nil layers are equally weighted, otherwise the length of
+// weights must equal the number of layers. If resolutions is nil, a resolution
+// of 1.0 is used for all layers, otherwise either a single element slice may be used
+// to specify a global resolution, or the length of resolutions must equal the number
+// of layers. The resolution parameter is γ as defined in Reichardt and Bornholdt
+// doi:10.1103/PhysRevE.74.016110.
+// QMultiplex will panic if the graph has any layer weight-scaled edge with
+// negative edge weight.
+//
+// If g is undirected, Q is calculated according to
+//  Q = \sum_{layer} w_{layer} \sum_{ij} [ A_{layer}*_{ij} - (\gamma_{layer} k_i k_j)/2m_{layer} ] \delta(c_i,c_j),
+// If g is directed, it is calculated according to
+//  Q = \sum_{layer} w_{layer} \sum_{ij} [ A_{layer}*_{ij} - (\gamma_{layer} k_i^in k_j^out)/m_{layer} ] \delta(c_i,c_j).
+//
+// Note that Q values for multiplex graphs are not scaled by the total layer edge weight.
+//
+// graph.Undirect may be used as a shim to allow calculation of Q for
+// directed graphs.
+func QMultiplex(g Multiplex, communities [][]graph.Node, weights, resolutions []float64) []float64 {
+	if weights != nil && len(weights) != g.Depth() {
+		panic("community: weights vector length mismatch")
+	}
+	if resolutions != nil && len(resolutions) != 1 && len(resolutions) != g.Depth() {
+		panic("community: resolutions vector length mismatch")
+	}
+
+	switch g := g.(type) {
+	case UndirectedMultiplex:
+		return qUndirectedMultiplex(g, communities, weights, resolutions)
+	case DirectedMultiplex:
+		return qDirectedMultiplex(g, communities, weights, resolutions)
+	default:
+		panic(fmt.Sprintf("community: invalid graph type: %T", g))
+	}
+}
+
+// ReducedMultiplex is a modularised multiplex graph.
+type ReducedMultiplex interface {
+	Multiplex
+
+	// Communities returns the community memberships
+	// of the nodes in the graph used to generate
+	// the reduced graph.
+	Communities() [][]graph.Node
+
+	// Structure returns the community structure of
+	// the current level of the module clustering.
+	// The first index of the returned value
+	// corresponds to the index of the nodes in the
+	// next higher level if it exists. The returned
+	// value should not be mutated.
+	Structure() [][]graph.Node
+
+	// Expanded returns the next lower level of the
+	// module clustering or nil if at the lowest level.
+	//
+	// The returned ReducedGraph will be the same
+	// concrete type as the receiver.
+	Expanded() ReducedMultiplex
+}
+
+// ModularizeMultiplex returns the hierarchical modularization of g at the given resolution
+// using the Louvain algorithm. If all is true and g have negatively weighted layers, all
+// communities will be searched during the modularization. If src is nil, rand.Intn is
```

Comment:
The `all` parameter says whether non-connected communities should be assessed as candidates to move a target node to. In the case of purely positively weighted multiplexes this is never useful, so we don't do it then. However, for negatively weighted multiplexes (or graphs), it may be the case that moving a node to a non-connected community is an improvement. This can be seen in the simple case of a single layer graph with only negative weights (yes, these are reasonable); here you can see that nodes that are connected to each other should not be placed in the same community. The repulsion test in `louvain_undirected_multplex_test.go` tests this.

For the second part, the definition of the multiplex Q looks like it shouldn't, but it seems to. The deltaQ for the multiplex case uses per layer m, which I think makes a difference.